### PR TITLE
Dataloader changes for RLVR

### PR DESCRIPTION
**Pull request by gupta-abhay**
- [x] Add script changes to write data
- [x] Add dataloader changes for verified answer
- [x] Add verifiable rewards (both defn and to registry)
- [x] Handle `str` vars for dataloader and callbacks, enabling path to more verifiers
- [x] Move verifier code to a single `rlvr_utils` file

### CODE REVIEW

**Review by vchiley — CHANGES_REQUESTED** [review]
Going to channel my inner Daniel: "add tests."

File: scripts/data/unified_tokenize_dataset.py [hunk 1]
```diff
@@ -128,6 +144,49 @@ def _process_classifier_sample(self, sample: Any):
             'label': np.asarray(label).tobytes(),
         }
 
+    def _get_processing_fn_from_dataset(self):
+        """Get the processing function based on the dataset name.
+
+        This function is currently hard-coded for the GSM8K dataset.
+        """
+        prompt_fn = prepare_gsm8k_prompt
```
Comment (SeanKski) [inline scripts/data/unified_tokenize_dataset.py:152]: maybe throw a log.warning here saying this this is hard-coded for GSM8K  (just to avoid a gotchya)

**Review by SeanKski — COMMENTED** [review]
Overall, I think this looks good. 

One quality-of-life thing that I think we should implement (but is not necessary now, so they shouldn't block this merge) is: 

  Rather than having an additional `<EvalName>FormatVerificationReward`, `<EvalName>AnswerVerificationReward` for every eval, we should build out some abstractions/standardizations, such as, a base `FormatVerificationReward` and `AnswerVerificationReward` which we then subclass/configure for the different evals. 
This way we won't overload the main `reward_learning` registry and cut down on repeated code 
e.g., otherwise, each outcome reward will have its own version of: 
```python 
       for i in range(batch_size):
            _answer = self._extract_solution(all_generated_texts[i])
            _reward = self._score_generations(_answer, verified_answers[i])
            rewards[i, generated_lens[i] - 1] += _reward
        return rewards
```

**Comment by gupta-abhay** [comment]
@SeanKski - addressed all your comments
@vchiley - added tests for rewards and for the dataloader workflow
@dakinggg - adding more dependencies - ptal if they look good - specifically the omegaconf installs which is required for antlr workflows.

File: pyproject.toml [hunk 2]
```diff
@@ -15,6 +15,10 @@ dependencies = [
     'ray<3',
     'kubernetes==29.0.0',
     'vllm>=0.8.2,<0.9.0',
+    # for MATH dataset
+    'omegaconf==2.4.0.dev3',
```
Comment (dakinggg) [inline pyproject.toml:19]: can you explain why you need exact pins for all of these libraries?

File: compose_rl/utils/utils.py [hunk 3]
```diff
@@ -1084,3 +1085,12 @@ def make_action_mask(
         mask[i, :prompt_len[i] - 1] = 0
 
     return mask
+
+
+def flatten(coll: Union[Iterable[Any], str]) -> Generator[Any, None, None]:
```
Comment (vchiley) [inline compose_rl/utils/utils.py:1090]: doc str
(cuz i'm pedantic)
use this one or do you

```suggestion
def flatten(coll: Union[Iterable[Any], str]) -> Generator[Any, None, None]:
    """Recursively flattens an arbitrarily nested iterable (excluding strings).

    Note: strings are treated as atomic elements and are not flattened into
    characters.

    Args:
        coll (Union[Iterable[Any], str]): The nested iterable to flatten.

    Yields:
        Any: The individual, non-iterable elements from the flattened structure.
    """
```

**Review by abaheti95 — APPROVED** [review]
Looks good. Thanks!
Probably need to remove one function that is unused.

File: pyproject.toml (hunk 2 above)
Comment (dakinggg) [reply pyproject.toml:19]: I skimmed the omegaconf commits between 2.3 and 2.4.0.dev3 and don't have a concern about that one. 

can you do a >=1.13.1,<2 for sympy?

where does the antlr dep come from? i.e. who uses it and do they have the dep pinned?

Comment (gupta-abhay) [reply pyproject.toml:19]: - The `antlr-runtime` and `sympy` library are sort of intertwined - they need to be fixed to versions, else it leads to weird parsing errors. 
- For `omegaconf`, if I do not specify the library, it defaults to the `2.3.0` - which does not suport `antlr-4.11` -- which is inturn needed for the parsing.

File: compose_rl/data/rlvr_utils.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        return False
+    except ImportError as e:
+        log.error(e)
+        raise
+    except Exception as e:
+        log.debug(f'Failed comparing {x1} and {x2} with {e}')
+        return False
+
+
+def get_unnormalized_answer(text: str) -> str:
```
[the remark this answers is not in the document]
Comment (gupta-abhay) [reply compose_rl/data/rlvr_utils.py:143]: ahh good catch, will remove

File: pyproject.toml (hunk 2 above)
Comment (dakinggg) [reply pyproject.toml:19]: ok i see, there is a dynamic version check in sympy when you call the parse latex stuff. Can you leave a comment behind explaining these things?

File: compose_rl/data/prompt_data.py [hunk 5]
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
...
-        return {'prompt': prompt, 'prompt_len': prompt_len}
+        item_dict = {'prompt': prompt, 'prompt_len': prompt_len}
+
+        verified_answer = sample.get('verified_answer', None)
+        if verified_answer:
+            if isinstance(verified_answer, str):
+                _answer = verified_answer
+            else:
+                try:
+                    _answer = verified_answer.decode('utf-8', errors='strict')
```
Comment (dakinggg) [inline compose_rl/data/prompt_data.py:109]: 😱 you have verified answers with non utf-8 decodable

File: compose_rl/data/prompt_data.py [hunk 6]
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
...
+                except UnicodeDecodeError:
+                    print(f'Failed to decode verifed_answer')
```
Comment (dakinggg) [inline compose_rl/data/prompt_data.py:111]: nit: log instead of print

File: compose_rl/data/rlvr_utils.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        left = '\\boxed{'
+        assert s[:len(left)] == left
+        return s[len(left):-1]
+
+    # Just remove any \boxed or \fbox prefix and any trailing brace
+    s = s.replace('\\boxed', '').replace('\\fbox', '')
+    return s.strip('{}')
+
+
+class timeout:  # noqa: N801
```
Comment (dakinggg) [inline compose_rl/data/rlvr_utils.py:75]: lol i asked claude whether this approach or the one in foundry (https://github.com/mosaicml/llm-foundry/blob/60e24b066d0037678de54c5ffaaf09b91ea9d8b7/llmfoundry/callbacks/run_timeout_callback.py#L4) works better, and the answer was that the one in foundry works on windows 😂

Comment (dakinggg) [reply compose_rl/data/rlvr_utils.py:75]: I suppose the one in foundry (threading.Timer) might be slightly better cause the signal.alarm calls here are process global right?

File: pyproject.toml (hunk 2 above)
Comment (dakinggg) [reply pyproject.toml:19]: inline here in pyproject.toml

File: compose_rl/data/prompt_data.py (hunk 6 above)
Comment (vchiley) [inline compose_rl/data/prompt_data.py:111]:
```suggestion
                except UnicodeDecodeError as e:
                    log.error(f'Failed to decode verifed_answer with error: {e}')
```

File: compose_rl/data/prompt_data.py [hunk 8]
```diff
@@ -90,5 +98,19 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
             prompt = prompt[:-truncate_len]
 
         prompt_len = torch.Tensor([len(prompt)]).to(dtype=torch.int64)
-
-        return {'prompt': prompt, 'prompt_len': prompt_len}
+        item_dict = {'prompt': prompt, 'prompt_len': prompt_len}
+
+        verified_answer = sample.get('verified_answer', None)
+        if verified_answer:
```
Comment (vchiley) [inline compose_rl/data/prompt_data.py:104]:
```suggestion
        if verified_answer is not None:
```
in case verified answer is an empty string (`''`) for some reason
or should this case be caught as a separate case and dealt with separately?

Comment (gupta-abhay) [reply compose_rl/data/prompt_data.py:104]: this is handled in the initial data processing script here: https://github.com/databricks/compose-rl/blob/abhay/data_loader_rlvr/scripts/data/unified_tokenize_dataset.py#L208

Comment (gupta-abhay) [reply compose_rl/data/prompt_data.py:104]: so we will not have empty strings here

File: compose_rl/data/prompt_data.py (hunk 5 above)
Comment (gupta-abhay) [reply compose_rl/data/prompt_data.py:109]: extra checks if any user makes weird datasets ❌

File: compose_rl/data/rlvr_utils.py (hunk 7 above)
Comment (gupta-abhay) [reply compose_rl/data/rlvr_utils.py:75]: yeah, you might be right 🤔 . For now, I will use this, since the minieval class uses the same functions - helps keep the evals and training closer to each other. We can mark this for future improvements as needed.

File: compose_rl/reward_learning/functional.py [hunk 9]
```diff
@@ -242,3 +250,190 @@ def __call__(
...
+    def __init__(self, cfg: dict[Any, Any], tokenizer: Tokenizer):
+        super().__init__(cfg, tokenizer)
+        self.reward = cfg.get('reward', 1.0)
+        log.info(
+            f'Using reward value of {self.reward} for {self.__class__.__name__} verifier',
+        )
+
+    def validate_config(self):
+        # Base validation, can be overridden
+        return
```
Comment (vchiley) [inline compose_rl/reward_learning/functional.py:268]: @bcui-db should validate_config be a member method of `Reward` cls since a bunch of these fn's define it?


Also
`pass` == `return` == `return None`, but `pass` is slightly more explicit for indicating a deliberate no-operation placeholder.
```suggestion
    def validate_config(self):
        # Base validation, can be overridden
        pass
```
maybe I'm just pedantic 🤷‍♂️

File: compose_rl/reward_learning/functional.py [hunk 10]
```diff
@@ -242,3 +250,190 @@ def __call__(
...
+        """Indicate that this verifier doesn't need extraction."""
+        return False
+
+    def score_generations(self, answer: str, label: str) -> float:
+        """Check if the answer follows the expected format with '####' marker.
+
+        Note: The label parameter is not used in this implementation but is required
+        by the interface.
+        """
+        solution = re.search(r'####.*?([\d,]+(?:\.\d+)?)', answer)
```
Comment (vchiley) [inline compose_rl/reward_learning/functional.py:400]: does this line belong in the `extract_solution` fn?
single line fn

File: scripts/data/unified_tokenize_dataset.py [hunk 11]
```diff
@@ -31,19 +39,25 @@ def __init__(
...
     ):
         self.tokenizer = tokenizer
         os.environ['TOKENIZERS_PARALLELISM'] = 'false'
         self.max_length = max_length
         self.dataset_type = dataset_type
+        self.dataset_name = dataset_name.lower()
 
         print(f'Dataset name: {dataset_name}')
+        if subset:
+            print(f'Processing subset: {subset}')
```
Comment (vchiley) [inline scripts/data/unified_tokenize_dataset.py:54]: this file has a bunch of prints
can you update these to `log` instead
🙏

File: tests/common/datasets.py [hunk 12]
```diff
@@ -72,3 +72,20 @@ def __getitem__(self, index: int):
...
+        self.size = size
+        self.prompt_len = prompt_len
+
+    def __len__(self):
+        return self.size
+
+    def __getitem__(self, index: int):
+        return {
+            'prompt': torch.ones((self.prompt_len,)).int(),
+            'prompt_len': torch.Tensor([self.prompt_len]).to(torch.int64),
```
Comment (vchiley) [inline tests/common/datasets.py:89]: torch.int64 needed?
why not `uint32`
goes from 0 to > 4B?

**Review by vchiley — APPROVED** [review]
left a few pedantic comments
but generally, lgtm

File: compose_rl/reward_learning/functional.py (hunk 10 above)
Comment (gupta-abhay) [reply compose_rl/reward_learning/functional.py:400]: can be done that way - but since we just need to check if the 4 hashes are followed, this simplifies stuff by putting it in a single function. else, it will complicate the interface with the `extract_solution` needing more fine grained handling of a None input. I dont think we currently have that diversified a need, and can let this be here.

File: compose_rl/reward_learning/functional.py [hunk 13]
```diff
@@ -242,3 +250,190 @@ def __call__(
...
+        """Score the generated answer against the label.
+
+        Args:
+            answer (str): The extracted answer.
+            label (str): The verified answer.
+
+        Returns:
+            float: The reward score.
+        """
+        pass
```
[the remark this answers is not in the document]
Comment (gupta-abhay) [reply compose_rl/reward_learning/functional.py:345]: since this is an abstractmethod - pass should be fine. Had it not been an abstract method - would necessitate the `raise` method.

File: tests/common/datasets.py (hunk 12 above)
Comment (gupta-abhay) [reply tests/common/datasets.py:89]: Keeps this the same as the actual code definition - which does int64 -- we can mark for future refinements as needed.

File: compose_rl/reward_learning/functional.py (hunk 9 above)
Comment (gupta-abhay) [reply compose_rl/reward_learning/functional.py:268]: @vchiley - noted. Will raise a separate PR for this. This Pr is already pretty large, and this will affect several files.
